<a href="https://colab.research.google.com/github/Eshashr/My-Projects/blob/main/DataMiningTechniquesProject1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project One

This project involves training two networks. The first is tackles a classification task and the second a regression 

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import altair as alt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from tensorflow import keras

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Part One - Binary  Classification Network

You will be training a binary classification notebook using the bank note authentication training set.

Look at the 'Binary Classification Learning Example' notebook to see how to setup a tensorflow neural network.

Create a network with 2 layers. The first layer should have 2 neurons, an input shape of (4, ) as there are 4 features, and use the swish activation function. The second layer should have 1 neuron as there is 1 output for binary classification and use the sigmoid activation function. 

Then compile the network using a learning rate of 0.05, the BinaryCrossEntropy() loss function, and the BinaryAccuracy() metric. 

Train the network for 50 epochs, as in the example use an early stopping callback.

### Questions
1. Run your training and evaluation code 3 times and record the number of training epochs needed, the final training accuracy and the final testing accuracy. Remember that with early stopping, your final training accuracy is at the point of lowest loss (likely 3 epochs back from when the training actually stopped).

2. Are the results reasonably consistent?

3. Are you seeing significant differences in accuracy between training and testing. What might this mean?

4. Increase the learning rate to 0.5 and re-record the results from 3 runs (as in question 1).

5. Are these results appreciably different from the earlier results on the first network?

Record your answer in a text cell.

In [ ]:
# Read in our classification dataset
df = pd.read_csv('/content/drive/MyDrive/HDS/Datasets/BankNoteAuthentication.csv')
df.head()

variance  skewness  curtosis  entropy  class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

In [ ]:
# Pull out features and targets
features = df.iloc[:, :-1].values
targets = df.iloc[:,-1].values

# Split off a test set
train_features, test_features, train_targets, test_targets = train_test_split(features, targets, test_size=0.2, random_state=42)

# Scale the features to the range 0-1
scaler = MinMaxScaler().fit(train_features)
train_features = scaler.transform(train_features)
test_features = scaler.transform(test_features)


Create a network with 2 layers. The first layer should have 2 neurons, an input shape of (4, ) as there are 4 features, and use the swish activation function. The second layer should have 1 neuron as there is 1 output for binary classification and use the sigmoid activation function.

Then compile the network using a learning rate of 0.05, the BinaryCrossEntropy() loss function, and the BinaryAccuracy() metric.

Train the network for 50 epochs, as in the example use an early stopping callback.

In [ ]:
# Put code in this cell and create new cells below for more code
# Our networks are always sequential
nn = keras.Sequential() # Create the network

# Add the input layer and specify its activation function
# We have 30 features so the input shape is (4, )
# Lets start with 2 neurons on this layer using swish
nn.add(keras.layers.Dense(2, input_shape=(4,), activation='swish'))

# Add the second layer and specify its activation function
# We want one output
# The output should be between 0 and 1
# so we use a sigmoid activation function
nn.add(keras.layers.Dense(1, activation='sigmoid'))

# Specify the learning algorithm (sometimes including learning rate),
# the loss (error) function, and metrics which are what
# do we want to know as the network is running
nn.compile(optimizer=keras.optimizers.RMSprop(0.05),
           loss=keras.losses.BinaryCrossentropy(),
           metrics=[keras.metrics.BinaryAccuracy()])

# Early stopping callback will halt training if loss increases
callback = keras.callbacks.EarlyStopping(monitor='loss', patience=3)

# Finally train the network
print("Training History:")
nn.fit(train_features, train_targets, epochs=50, callbacks=[callback])
# Evaluate its performance on the test set
print("Testing Results:")
nn.evaluate(test_features, test_targets)

Training History:
Epoch 1/50
35/35 [==============================] - 1s 1ms/step - loss: 0.6282 - binary_accuracy: 0.6272
Epoch 2/50
35/35 [==============================] - 0s 1ms/step - loss: 0.4200 - binary_accuracy: 0.8140
Epoch 3/50
35/35 [==============================] - 0s 2ms/step - loss: 0.2809 - binary_accuracy: 0.9070
Epoch 4/50
35/35 [==============================] - 0s 1ms/step - loss: 0.1967 - binary_accuracy: 0.9490
Epoch 5/50
35/35 [==============================] - 0s 2ms/step - loss: 0.1537 - binary_accuracy: 0.9535
Epoch 6/50
35/35 [==============================] - 0s 2ms/step - loss: 0.1267 - binary_accuracy: 0.9617
Epoch 7/50
35/35 [==============================] - 0s 2ms/step - loss: 0.1102 - binary_accuracy: 0.9581
Epoch 8/50
35/35 [==============================] - 0s 2ms/step - loss: 0.1039 - binary_accuracy: 0.9644
Epoch 9/50
35/35 [==============================] - 0s 2ms/step - loss: 0.0891 - binary_accuracy: 0.9699
Epoch 10/50
35/35 [==================

[0.06253109127283096, 0.9781818389892578]

1. Run your training and evaluation code 3 times and record the number of training epochs needed, the final training accuracy and the final testing accuracy. Remember that with early stopping, your final training accuracy is at the point of lowest loss (likely 3 epochs back from when the training actually stopped).

1st time:
Number of Ephochs: 21/50 epochs,  loss: 0.0363 - binary_accuracy: 0.9863

The final training accuracy: 0.044893886893987656,  

The final testing accuracy: 0.9781818389892578

2nd Time:
Number of Ephochs: 23/50 epochs, loss: 0.0530 - binary_accuracy: 0.9836

The final training accuracy: 0.07382497936487198,

The final testing accuracy:  0.9745454788208008

3rd time:
Number of Ephochs:27/50 epochs, loss: 0.0408 - binary_accuracy: 0.9809

The final training accuracy: 0.06253109127283096, 

The final testing accuracy:0.9781818389892578

2. Are the results reasonably consistent?
Yes, the results are reasonably consistent.

3. Are you seeing significant differences in accuracy between training and testing. What might this mean?
There is no significant difference.


In [ ]:
# Our networks are always sequential
nn = keras.Sequential() # Create the network

# Add the input layer and specify its activation function
# We have 30 features so the input shape is (4, )
# Lets start with 2 neurons on this layer using swish
nn.add(keras.layers.Dense(2, input_shape=(4,), activation='swish'))

# Add the second layer and specify its activation function
# We want one output
# The output should be between 0 and 1
# so we use a sigmoid activation function
nn.add(keras.layers.Dense(1, activation='sigmoid'))

# Specify the learning algorithm (sometimes including learning rate),
# the loss (error) function, and metrics which are what
# do we want to know as the network is running
nn.compile(optimizer=keras.optimizers.RMSprop(0.5),
           loss=keras.losses.BinaryCrossentropy(),
           metrics=[keras.metrics.BinaryAccuracy()])

# Early stopping callback will halt training if loss increases
callback = keras.callbacks.EarlyStopping(monitor='loss', patience=3)

# Finally train the network
print("Training History:")
nn.fit(train_features, train_targets, epochs=50, callbacks=[callback])
# Evaluate its performance on the test set
print("Testing Results:")
nn.evaluate(test_features, test_targets)

Training History:
Epoch 1/50
35/35 [==============================] - 1s 1ms/step - loss: 0.7094 - binary_accuracy: 0.5488
Epoch 2/50
35/35 [==============================] - 0s 1ms/step - loss: 0.6782 - binary_accuracy: 0.6026
Epoch 3/50
35/35 [==============================] - 0s 2ms/step - loss: 0.5669 - binary_accuracy: 0.6928
Epoch 4/50
35/35 [==============================] - 0s 1ms/step - loss: 0.5562 - binary_accuracy: 0.8241
Epoch 5/50
35/35 [==============================] - 0s 1ms/step - loss: 0.4442 - binary_accuracy: 0.8122
Epoch 6/50
35/35 [==============================] - 0s 1ms/step - loss: 0.4212 - binary_accuracy: 0.8396
Epoch 7/50
35/35 [==============================] - 0s 2ms/step - loss: 0.3521 - binary_accuracy: 0.8350
Epoch 8/50
35/35 [==============================] - 0s 1ms/step - loss: 0.4480 - binary_accuracy: 0.8669
Epoch 9/50
35/35 [==============================] - 0s 2ms/step - loss: 0.3895 - binary_accuracy: 0.8733
Epoch 10/50
35/35 [==================

[0.1353023499250412, 0.9781818389892578]

4. Increase the learning rate to 0.5 and re-record the results from 3 runs (as in question 1).
1st time:
Number of Ephochs: 9/50 epochs,  loss: 0.1195 - binary_accuracy: 0.9599

The final training accuracy: 0.08231337368488312, 

The final testing accuracy: 0.9672726988792419

2nd Time:
Number of Ephochs: 11/50 epochs, loss: 0.1312 - binary_accuracy: 0.9562

The final training accuracy: 0.15099488198757172,  

The final testing accuracy: 0.9527272582054138

3rd time:
Number of Ephochs:12/50 epochs, loss: 0.2140 - binary_accuracy: 0.9471

The final training accuracy: 0.1353023499250412, 

The final testing accuracy: 0.9781818389892578

5. Are these results appreciably different from the earlier results on the first network?
There are still huge differences between the training rate and testing rate. However, it only required fewer epochs to get the final result.
Record your answer in a text cell.

## Part Two - Regression Network

You will be training a network to learn the function $sin(x)cos(x)$.

Look at the 'Regression Learning Example' notebook to see how to setup a tensorflow neural network for this type of task.

Create a network with 3 layers. The first layer should have 8 neurons, an input shape of (1, ) as there is 1 feature, and use the swish activation function. The second layer should have 4 neurons and also use swish. The last layer should have 1 neuron as there is 1 output for the regression and use the linear activation function. 

Then compile the network using a learning rate of 0.05, the MeanSquaredError() loss function, and the MeanAbsoluteError() metric. 

Train the network for 50 epochs, as in the example use an early stopping callback.

### Questions
1. Run your training and evaluation code 3 times and record the number of training epochs needed, the final training accuracy and the final testing accuracy. Remember that with early stopping, your final training accuracy is at the point of lowest loss (likely 3 epochs back from when the training actually stopped).

2. Are the results reasonably consistent?

3. Are you seeing significant differences in accuracy between training and testing. What might this mean?

4. Create a plot of the actual output of one of your runs compared to the target output. An example of such a plot is in the Regression Learning Example notebook.

5. Copy your training code to a new cell. Increase the first and second layers to 64 neurons and re-record the results from 3 runs (as in question 1).

6. Are these results appreciably different from the earlier results on the first network? If so, why might that be the case?

7. Again create a plot of the actual output of one of your runs with this new larger network compared to the target output.

Record your answers in text cells.

In [ ]:
# Create the dataset
x = np.linspace(0,2*np.pi,1001)
y = np.sin(x)*np.cos(x)
df = pd.DataFrame({"Input":x, "Target":y})
alt.Chart(df, title="cos(x)*sin(x)").mark_line().encode(
    x='Input:Q',
    y='Target:Q'
)

alt.Chart(...)

In [ ]:
# Pull out features and targets
features = df.iloc[:, :-1].values
targets = df.iloc[:,-1].values

# Split off a test set
train_features, test_features, train_targets, test_targets = train_test_split(features, targets, test_size=0.2, random_state=42)

# Skip scaling as their is only one feature and its is in a reasonable range

Create a network with 3 layers. The first layer should have 8 neurons, an input shape of (1, ) as there is 1 feature, and use the swish activation function. The second layer should have 4 neurons and also use swish. The last layer should have 1 neuron as there is 1 output for the regression and use the linear activation function.

Then compile the network using a learning rate of 0.05, the MeanSquaredError() loss function, and the MeanAbsoluteError() metric.

Train the network for 50 epochs, as in the example use an early stopping callback.

In [ ]:
# Put code in this cell and create new cells below for more code
# Our networks are always sequential
nn = keras.Sequential() # Create the network

# Add the input layer and specify its activation function
# We have 1 feature so the input shape is (1, )
# Lets start with 8 neurons using swish
nn.add(keras.layers.Dense(8, input_shape=(1,), activation='swish'))

# Then add another hidden layer with 4 neurons
nn.add(keras.layers.Dense(4, activation='swish'))

# Add the output layer and specify its activation function
# We want one output with no limits on the range
# so we use a linear activation function
nn.add(keras.layers.Dense(1, activation='linear'))

# Specify the learning algorithm, loss (error) function and what
# do we want to know as the network is running
nn.compile(optimizer=keras.optimizers.RMSprop(learning_rate=0.05),
           loss=keras.losses.MeanSquaredError(),
           metrics=[keras.metrics.MeanAbsoluteError()])

# Early stopping callback will halt training if loss increases
callback = keras.callbacks.EarlyStopping(monitor='loss', patience=3)

# Finally train the network
print("Training History:")
nn.fit(train_features, train_targets, epochs=50, callbacks=[callback])
# Evaluate the results
print("Testing Results:")
nn.evaluate(test_features, test_targets)

Training History:
Epoch 1/50
25/25 [==============================] - 1s 1ms/step - loss: 0.5556 - mean_absolute_error: 0.4443
Epoch 2/50
25/25 [==============================] - 0s 1ms/step - loss: 0.1142 - mean_absolute_error: 0.2906
Epoch 3/50
25/25 [==============================] - 0s 1ms/step - loss: 0.1137 - mean_absolute_error: 0.2924
Epoch 4/50
25/25 [==============================] - 0s 2ms/step - loss: 0.1093 - mean_absolute_error: 0.2868
Epoch 5/50
25/25 [==============================] - 0s 2ms/step - loss: 0.1077 - mean_absolute_error: 0.2848
Epoch 6/50
25/25 [==============================] - 0s 2ms/step - loss: 0.0849 - mean_absolute_error: 0.2431
Epoch 7/50
25/25 [==============================] - 0s 1ms/step - loss: 0.0786 - mean_absolute_error: 0.2348
Epoch 8/50
25/25 [==============================] - 0s 2ms/step - loss: 0.0608 - mean_absolute_error: 0.2016
Epoch 9/50
25/25 [==============================] - 0s 2ms/step - loss: 0.0535 - mean_absolute_error: 0.1865
E

[0.02713038958609104, 0.14073190093040466]

1. Run your training and evaluation code 3 times and record the number of training epochs needed, the final training accuracy and the final testing accuracy. Remember that with early stopping, your final training accuracy is at the point of lowest loss (likely 3 epochs back from when the training actually stopped).

1st time:

Number of Ephochs: 13/50 epochs,  loss: 0.0749 - mean_absolute_error: 0.2201

The final training accuracy: 0.07776641100645065,  

The final testing accuracy: 0.23011787235736847

2nd Time:
Number of Ephochs: 9/50 epochs, loss: 0.0368 - mean_absolute_error: 0.1567

The final training accuracy: 0.056101057678461075, 

The final testing accuracy: 0.2047019749879837 

3rd time:
Number of Ephochs:14/50 epochs, loss: 0.0329 - mean_absolute_error: 0.1430

The final training accuracy: 0.03709542006254196, 

The final testing accuracy: 0.15053647756576538

2. Are the results reasonably consistent?
Yes these results are reasonably consistent.

3. Are you seeing significant differences in accuracy between training and testing. What might this mean?
In all three times, the Training accuracy < Testing Accuracy so underfit. It wasnt as much so my network was working properly.

4. Create a plot of the actual output of one of your runs compared to the target output. An example of such a plot is in the Regression Learning Example notebook.

5. Copy your training code to a new cell. Increase the first and second layers to 64 neurons and re-record the results from 3 runs (as in question 1).

6. Are these results appreciably different from the earlier results on the first network? If so, why might that be the case?

7. Again create a plot of the actual output of one of your runs with this new larger network compared to the target output.

Record your answers in text cells.

In [ ]:
# Look at the predictions
actual = nn.predict(x).flatten()
df["Actual"] = actual
alt.Chart(df, title="cos(x)*sin(x)").transform_fold(
    ['Target', 'Actual'],
    as_=['Result', 'Output']
).mark_line().encode(
    x='Input:Q',
    y='Output:Q',
    color='Result:N'
)

alt.Chart(...)

In [ ]:
# Create the dataset
x = np.linspace(0,2*np.pi,1001)
y = np.sin(x)*np.cos(x)
df = pd.DataFrame({"Input":x, "Target":y})
alt.Chart(df, title="cos(x)*sin(x)").mark_line().encode(
    x='Input:Q',
    y='Target:Q'
)

alt.Chart(...)

In [ ]:
features = df.iloc[:, :-1].values
targets = df.iloc[:,-1].values

# Split off a test set
train_features, test_features, train_targets, test_targets = train_test_split(features, targets, test_size=0.2, random_state=42)

In [ ]:
# Put code in this cell and create new cells below for more code
# Our networks are always sequential
nn = keras.Sequential() # Create the network

# Add the input layer and specify its activation function
# We have 1 feature so the input shape is (1, )
# Lets start with 8 neurons using swish
nn.add(keras.layers.Dense(64, input_shape=(1,), activation='swish'))

# Then add another hidden layer with 4 neurons
nn.add(keras.layers.Dense(64, activation='swish'))

# Add the output layer and specify its activation function
# We want one output with no limits on the range
# so we use a linear activation function
nn.add(keras.layers.Dense(1, activation='linear'))

# Specify the learning algorithm, loss (error) function and what
# do we want to know as the network is running
nn.compile(optimizer=keras.optimizers.RMSprop(learning_rate=0.05),
           loss=keras.losses.MeanSquaredError(),
           metrics=[keras.metrics.MeanAbsoluteError()])

# Early stopping callback will halt training if loss increases
callback = keras.callbacks.EarlyStopping(monitor='loss', patience=3)

# Finally train the network
print("Training History:")
nn.fit(train_features, train_targets, epochs=50, callbacks=[callback])
# Evaluate the results
print("Testing Results:")
nn.evaluate(test_features, test_targets)

Training History:
Epoch 1/50
25/25 [==============================] - 1s 1ms/step - loss: 30.6054 - mean_absolute_error: 2.3097
Epoch 2/50
25/25 [==============================] - 0s 2ms/step - loss: 0.1079 - mean_absolute_error: 0.2753
Epoch 3/50
25/25 [==============================] - 0s 2ms/step - loss: 0.2561 - mean_absolute_error: 0.3779
Epoch 4/50
25/25 [==============================] - 0s 2ms/step - loss: 0.9568 - mean_absolute_error: 0.5416
Epoch 5/50
25/25 [==============================] - 0s 2ms/step - loss: 0.1196 - mean_absolute_error: 0.2972
Testing Results:
7/7 [==============================] - 0s 3ms/step - loss: 0.0932 - mean_absolute_error: 0.2581


[0.09321625530719757, 0.2581304609775543]

5. Copy your training code to a new cell. Increase the first and second layers to 64 neurons and re-record the results from 3 runs (as in question 1).

1st time:

Number of Ephochs: 2/50 epochs,  loss: 0.1079 - mean_absolute_error: 0.2753

The final training accuracy: 0.09321625530719757, 

The final testing accuracy:   0.2581304609775543,

2nd Time:
Number of Ephochs: 2/50 epochs, loss: 0.1307 - mean_absolute_error: 0.3076

The final training accuracy: 0.127406045794487, 

The final testing accuracy: 0.28035128116607666 

3rd time:
Number of Ephochs:3/50 epochs, loss: 0.0994 - mean_absolute_error: 0.2653

The final training accuracy: 0.0837436318397522, 

The final testing accuracy: 0.23880192637443542

6. Are these results appreciably different from the earlier results on the first network? If so, why might that be the case?
The number of epochs needed were very few in this network. The error was very high and this might be because the nurons were increased to 64.

In [ ]:
# Look at the predictions
actual = nn.predict(x).flatten()
df["Actual"] = actual
alt.Chart(df, title="cos(x)*sin(x)").transform_fold(
    ['Target', 'Actual'],
    as_=['Result', 'Output']
).mark_line().encode(
    x='Input:Q',
    y='Output:Q',
    color='Result:N'
)

alt.Chart(...)

The target and the output were different.